### ISSUE TITLE
Autoscaler fails with HTTP 403 when querying Nomad /nodes API

### ISSUE THREAD

**Issue by jokramer123** [hashicorp/nomad-autoscaler#380]
We are currently using the autoscaler as a Nomad job that runs inside a **namespace** and has a scaling policy that references an AWS autoscaling group for a target. The autoscaler job is using an IAM role and retrieves a Nomad ACL token from Vault. The autoscaler detects our metric and scales up the aws autoscaling group target as expected. Afterwards it fails to scale down the aws autoscaling group and we see HTTP 403 errors when it attempts to hit the /nodes API. If I take the same token the autoscaler has I can curl the /nodes API successfully. This is blocking us from fully adopting the autoscaler and feels like a namespace problem.

- nomad enterprise v1.0.2.
- autoscaler v0.2.1 

Errors:

`2021-02-11T02:47:24.458Z [ERROR] policy_eval.worker.check_handler: failed to submit scaling action to target: check=batch_job_queue_size id=c90f4d86-1485-df13-48c2-6e3bb75fd11b policy_id=5514dc5e-445d-6aae-c54c-bb629f7a1d5d queue=cluster source=datadog strategy=target-value target=aws-asg error="failed to perform scaling action: failed to perform pre-scale Nomad scale in tasks: failed to identify nodes for removal: failed to list Nomad nodes from API: Unexpected response code: 403 (Permission denied)"`

`2021-02-11T02:47:24.458Z [ERROR] policy_eval.worker: failed to evaluate policy: eval_id=4da7f9b8-91cc-3ba5-eb93-8c3634521ee4 eval_token=6c67961f-3424-c06c-a71d-7281276cec7e id=c90f4d86-1485-df13-48c2-6e3bb75fd11b policy_id=5514dc5e-445d-6aae-c54c-bb629f7a1d5d queue=cluster err="failed to scale target: failed to perform scaling action: failed to perform pre-scale Nomad scale in tasks: failed to identify nodes for removal: failed to list Nomad nodes from API: Unexpected response code: 403 (Permission denied)"`

Any help here would be appreciated!

**Comment by jrasell** [comment]
Hi @jokramer123 and apologies you have ran into this. Would you be able to share (redacted as needed) the ACL policy you are using to create the ACL token that the autoscaler uses?

**Comment by jokramer123** [comment]
Hello @jrasell and thanks for the reply.

Here is the ACL policy that I have currently (I've tried a few variations)

`
node {
  policy = "read"
}

namespace "${namespace}" {
  policy = "scale"
}
`

**Comment by jrasell** [comment]
@jokramer123 thanks and whilst this doesn't solve the initial issue you're reporting here (read should allow listing of nodes), in order to fully run the scale in actions, the policy will need `node { policy = "write" }` due to the fact it triggers a node drain before terminating the remote resource.

In order to help reproduce this, would you be able to provide (again redacted) the autoscaler configuration you're using?

**Comment by jokramer123** [comment]
Thanks @jrasell the node 'write' makes sense, I updated the ACL policy and still get the same HTTP 403 however. 

Here is the autoscaler config:

```
policy {
  dir = "local/policies/"
  default_cooldown = "2m"
}

plugin_dir = "/plugins"

nomad {
  address = "$NOMAD_ADDR"
}
apm "nomad" {
  driver = "nomad-apm"
  config  = {
    address = "$NOMAD_ADDR"
  }
}

target "aws-asg" {
  driver = "aws-asg"
  config = {
    aws_region = "${region}"
  }
}

apm "datadog"{
   driver = "datadog"
   config = {
 {{ with secret "kv/data/shared"}}
     dd_api_key = "{{.Data.data.DATADOG_API_KEY}}"
     dd_app_key = "{{.Data.data.DATADOG_APP_KEY}}"
 {{ end }}
   }
}

strategy "target-value" {
  driver = "target-value"
}
          EOF

        destination = "local/config.hcl"
```
        
        
Here is my scaling policy:

```
scaling "cluster_policy" {
  enabled = true
  min     = 0
  max     = 5

  policy {
    cooldown            = "4m"
    evaluation_interval = "1m"

    check "batch_job_queue_size" {
      source = "datadog"
      query = "sum:nomad.nomad.job_summary.queued{namespace:${namespace}} by {namespace}"
      query_window = "4m"
      strategy "target-value" {
        target = 0
      }
    }

    target "aws-asg" {
      dry-run             = "false"
      aws_asg_name        = "${asg_name}"
      node_class          = "${namespace}"
      node_purge          = "true"
    }
  }
}
```

**Comment by jrasell** [comment]
Thanks @jokramer123. I don't see the Nomad ACL token being passed to the autoscaler config in the block you pasted, is this something that is missing, provided a different way, or has been removed for redaction purposes?

**Comment by jokramer123** [comment]
I have a terraform file which templates out the nomad job file and deploys it, here is the template. The NOMAD_TOKEN is part of this. To troubleshoot, I was taking this env var out of the running container and then using it to issue curl requests to the /nodes API myself which returns HTTP 200.

```

job "${asg_name}-autoscaler" {

  constraint {
    attribute = "$${node.class}"
    value = "${namespace}"
  }

  namespace = "${namespace}"
  datacenters = ["redacted"]

  group "autoscaler" {
    count = 1

    task "autoscaler" {
      template {
      data = <<EOH
      NOMAD_TOKEN="{{with secret "nomad/creds/${namespace}-autoscaler-${asg_name}"}}{{.Data.secret_id}}{{end}}"
      EOH

      destination = "secrets/file.env"
      env         = true
      }

      vault {
        policies = [
          "${namespace}-nomad-client-policy"
        ]
      }

      driver = "docker"

      config {
        image   = "hashicorp/nomad-autoscaler:0.2.1"
        command = "nomad-autoscaler"
        args    = [
          "agent",
          "-config",
          "$${NOMAD_TASK_DIR}/config.hcl",
          "-log-level",
          "debug"
        ]
        port_map {
          http = 8080
        }
      }

# ---------------------------------------------------------------------------------------------------------------------
# Autoscaler config
# This template stanza is for the autoscaler config
# ---------------------------------------------------------------------------------------------------------------------

      template {
        data = <<EOF

policy {
  dir = "local/policies/"
  default_cooldown = "2m"
}

plugin_dir = "/plugins"

nomad {
  address = "$NOMAD_ADDR"
}
apm "nomad" {
  driver = "nomad-apm"
  config  = {
    address = "$NOMAD_ADDR"
  }
}

target "aws-asg" {
  driver = "aws-asg"
  config = {
    aws_region = "${region}"
  }
}

apm "datadog"{
   driver = "datadog"
   config = {
 {{ with secret "kv/data/shared"}}
     dd_api_key = "{{.Data.data.DATADOG_API_KEY}}"
     dd_app_key = "{{.Data.data.DATADOG_APP_KEY}}"
 {{ end }}
   }
}

strategy "target-value" {
  driver = "target-value"
}
          EOF

        destination = "local/config.hcl"
      }

# ---------------------------------------------------------------------------------------------------------------------
# Autoscaler policy
# ---------------------------------------------------------------------------------------------------------------------

template {
        data = <<EOF

scaling "cluster_policy" {
  enabled = true
  min     = 0
  max     = 5

  policy {
    cooldown            = "4m"
    evaluation_interval = "1m"

    check "batch_job_queue_size" {
      source = "datadog"
      query = "sum:nomad.nomad.job_summary.queued{namespace:${namespace}} by {namespace}"
      query_window = "4m"
      strategy "target-value" {
        target = 0
      }
    }

    target "aws-asg" {
      dry-run             = "false"
      aws_asg_name        = "${asg_name}"
      node_class          = "${namespace}"
      node_purge          = "true"
    }
  }
}
          EOF
        destination = "local/policies/batch.hcl"
      }

    resources {
      cpu    = 300
      memory = 512

      network {
        mbits = 10
        port "http" {}
      }
    }

    service {
      name = "autoscaler"
      port = "http"

      check {
        type     = "http"
        path     = "/v1/health"
        interval = "5s"
        timeout  = "4s"
        }
      }
    }
  }
}
```

**Comment by jrasell** [comment]
Thanks for all the detail @jokramer123, I found the problem and have raised a fix within #381. Once this is reviewed we will be able to merged this into the main branch.
